### PR TITLE
Add feedforward data to combined diagnostic plots

### LINKED ISSUE #107: Display feedforward analysis fields and slider positions in combined diagnostics window
The motivation behind this is providing enough context to reproduce the plots shown in the screenshot without having to also screenshot the background window.

The fields and sliders could be read-only or read-write, whatever's easier to implement.

### PR DESCRIPTION
**Pull request by Piphi5**
Mainly allows users to also screenshot their settings + gains whenever sharing graphs.
Fixes #107 
image

### CODE REVIEW

**Comment by calcmogul** [comment]
How close is this code to the feedforward analysis dropdown? Could they be refactored into a single function called from the dropdown as well as the combined diagnostics popup?

File: src/main/native/cpp/view/Analyzer.cpp [hunk 1]
```diff
@@ -41,6 +42,17 @@ Analyzer::Analyzer(wpi::Logger& logger) : m_logger(logger) {
...
+auto ShowGain = [](const char* text, double* data) {
+  ImGui::SetNextItemWidth(ImGui::GetFontSize() * 5);
+  ImGui::InputDouble(text, data, 0.0, 0.0, "%.5G",
+                     ImGuiInputTextFlags_ReadOnly);
+};
+
+auto SetPosition = [](double beginX, double beginY, int xShift, int yShift) {
+  ImGui::SetCursorPos(ImVec2(beginX + xShift * 10 * ImGui::GetFontSize(),
+                             beginY + yShift * 1.75 * ImGui::GetFontSize()));
+};
```
Comment (calcmogul) [inline src/main/native/cpp/view/Analyzer.cpp:54]: These should just be static functions if they're in global scope.

**Review by calcmogul — CHANGES_REQUESTED** [review]

**Comment by calcmogul** [comment]
The test duration is ambiguous. It looks more like "1.43/3 seconds" instead of "1.43 seconds out of a possible 3 seconds".

**Comment by Piphi5** [comment]
would something like `x` out of `total` make it clearer?

**Comment by calcmogul** [comment]
Yea

File: src/main/native/cpp/view/Analyzer.cpp [hunk 2]
```diff
@@ -41,6 +42,17 @@ Analyzer::Analyzer(wpi::Logger& logger) : m_logger(logger) {
...
+static auto ShowGain = [](const char* text, double* data) {
+  ImGui::SetNextItemWidth(ImGui::GetFontSize() * 5);
+  ImGui::InputDouble(text, data, 0.0, 0.0, "%.5G",
+                     ImGuiInputTextFlags_ReadOnly);
+};
+
+static auto SetPosition = [](double beginX, double beginY, int xShift, int yShift) {
+  ImGui::SetCursorPos(ImVec2(beginX + xShift * 10 * ImGui::GetFontSize(),
+                             beginY + yShift * 1.75 * ImGui::GetFontSize()));
+};
```
Comment (calcmogul) [inline src/main/native/cpp/view/Analyzer.cpp:54]:
```suggestion
static void ShowGain(const char* text, double* data) {
  ImGui::SetNextItemWidth(ImGui::GetFontSize() * 5);
  ImGui::InputDouble(text, data, 0.0, 0.0, "%.5G",
                     ImGuiInputTextFlags_ReadOnly);
}

static void SetPosition(double beginX, double beginY, int xShift, int yShift) {
  ImGui::SetCursorPos(ImVec2(beginX + xShift * 10 * ImGui::GetFontSize(),
                             beginY + yShift * 1.75 * ImGui::GetFontSize()));
}
```

**Review by calcmogul — CHANGES_REQUESTED** [review]

**Comment by Piphi5** [comment]
@calcmogul 
Is this clearer?
image

File: src/main/native/cpp/view/Analyzer.cpp [hunk 3]
```diff
@@ -41,6 +42,18 @@ Analyzer::Analyzer(wpi::Logger& logger) : m_logger(logger) {
...
+  ImGui::SetNextItemWidth(ImGui::GetFontSize() * 5);
+  ImGui::InputDouble(text, data, 0.0, 0.0, "%.5G",
+                     ImGuiInputTextFlags_ReadOnly);
+};
+
+static auto SetPosition = [](double beginX, double beginY, int xShift,
+                             int yShift) {
+  ImGui::SetCursorPos(ImVec2(beginX + xShift * 10 * ImGui::GetFontSize(),
+                             beginY + yShift * 1.75 * ImGui::GetFontSize()));
+};
```
Comment (calcmogul) [inline src/main/native/cpp/view/Analyzer.cpp:55]: Use normal functions here, not lambdas.

**Comment by calcmogul** [comment]
The unit "(s)" should be part of the label instead. Otherwise, the UI looks good. Only other thing is the static lambdas should be plain old functions.

**Comment by Piphi5** [comment]
image

File: src/main/native/cpp/view/Analyzer.cpp [hunk 4]
```diff
@@ -587,3 +536,88 @@ void Analyzer::ResetManagerState() {
...
+                           m_manager->GetMinDuration(),
+                           m_manager->GetMaxDuration(), "%.2f")) {
+      m_settings.stepTestDuration = units::second_t{m_stepTestDuration};
+      RefreshInformation();
+    }
+
+    ImGui::SetCursorPosY(endY);
+  } else {
+    std::ostringstream ss;
+    ss << std::round(m_stepTestDuration * 100) / 100 << " of "
```
Comment (prateekma) [inline src/main/native/cpp/view/Analyzer.cpp:614]: Use `wpi::format` from `<wpi/Format.h>` instead.

**Review by prateekma — CHANGES_REQUESTED** [review]

File: src/main/native/cpp/view/Analyzer.cpp (hunk 4 above)
Comment (Piphi5) [reply src/main/native/cpp/view/Analyzer.cpp:614]: Is that similar to the c++20 std::format?

Comment (calcmogul) [reply src/main/native/cpp/view/Analyzer.cpp:614]: Not really. fmtlib is more like C++20 std::format.

Comment (prateekma) [reply src/main/native/cpp/view/Analyzer.cpp:614]: You can do `wpi::format(v, "%.2f")`

Comment (Piphi5) [reply src/main/native/cpp/view/Analyzer.cpp]: Is it properly implemented?

File: src/main/native/include/sysid/view/Analyzer.h [hunk 5]
```diff
@@ -51,6 +51,7 @@ class Analyzer : public glass::View {
   void PrepareGraphs();
   void RefreshInformation();
   void AbortDataPrep();
+  void LoadFeedForwardGains(bool combined = false);
```
Comment (prateekma) [inline src/main/native/include/sysid/view/Analyzer.h:54]: Can you rename this to `DisplayFeedforwardGains()`? Feedforward is also one word.

**Review by prateekma — CHANGES_REQUESTED** [review]

File: src/main/native/cpp/view/Analyzer.cpp [hunk 6]
```diff
@@ -6,6 +6,7 @@
 
 #include <algorithm>
 #include <exception>
+#include <sstream>
```
Comment (calcmogul) [inline src/main/native/cpp/view/Analyzer.cpp:9]: Remove this unused include.

**Review by calcmogul — CHANGES_REQUESTED** [review]

File: src/main/native/include/sysid/view/Analyzer.h [hunk 7]
```diff
@@ -51,6 +51,7 @@ class Analyzer : public glass::View {
   void PrepareGraphs();
   void RefreshInformation();
   void AbortDataPrep();
+  void LoadFeedforwardGains(bool combined = false);
```
Comment (calcmogul) [inline src/main/native/include/sysid/view/Analyzer.h:54]: It's not loading anything, just displaying it.
```suggestion
  void DisplayFeedforwardGains(bool combined = false);
```

Comment (Piphi5) [reply src/main/native/include/sysid/view/Analyzer.h:54]: Yeah that makes more sense

**Review by calcmogul — APPROVED** [review]

**Review by prateekma — APPROVED** [review]

**Pull request merged**